### PR TITLE
Feature progress reports #86

### PR DESCRIPTION
**Pull request by adamtlord**
##### Feature list
**Django**
* Added a fk from ProgressReport to ProgrammeDocument. They are connected in a more complicated way also, but because every ProgressReport must be associated with a ProgrammeDocument, this seemed like a good shortcut that would serve us well in the future.
* Added a new API endpoint in the unicef app that returns a list of all Progress Reports -- _not currently filtered by location_. This may need to change.
* Added filterset for this endpoint.
* Added unit tests for this view

**Polymer**
* Added a new item to the left nav, and a new page, `progress-reports`
* This page contains two components, `progress-reports-list` and `progress-reports-filters`
* Created a new filter component, `date-filter`, incorporating [`etools-datepicker`](https://github.com/unicef-polymer/etools-datepicker) and borrowing liberally from the [`etools-date-input`](https://github.com/unicef/etools-partnership-management/blob/master/pmp/app-elements/common-elements/etools-date-input.html) component in PMP. This also requried a new set of `DateBehavior`s.
* `date-filter` required the addition of some bower components, notably [moment-element](https://github.com/bendavis78/moment-element) for js date manipulation.

**Note:** Currently the checkbox-filter for "Show only due reports" adds another GET param, "due", a Boolean. This is really just filtering ProgressReports by `status__in=[due, over_due]`. I think it would make more sense for this to also use the status filter, but I couldn't figure out how to make `iron-query-params` work with Django's MultipleChoiceFilter (eg, &status=due,over_due vs &status=due&status=over_due)

##### Progress checker
- [x] Django

- [x] Polymer

- [x] Django test cases

- [x] Polymer test cases

### CODE REVIEW

File: django_api/django_api/apps/unicef/views.py [hunk 1]
```diff
@@ -79,3 +81,16 @@ def get_object(self, pk):
             return ProgrammeDocument.objects.get(pk=pk)
         except ProgrammeDocument.DoesNotExist:
             raise Http404
+
+
+class ProgressReportAPIView(ListAPIView):
+    """
+    Endpoint for getting list of all PD Progress Reports
+    """
+    queryset = ProgressReport.objects.all()
```
Comment (lkolacz) [inline django_api/django_api/apps/unicef/views.py]: I'm afraid that this want be so simple ...
@n1207n Shouldn't it be something  like this where location id is a chosen location from top-right drop down menu:
```
    def get_queryset(self):
        pd_ids = Location.objects.filter(
            Q(id=self.location_id) |
            Q(parent_id=self.location_id) |
            Q(parent__parent_id=self.location_id) |
            Q(parent__parent__parent_id=self.location_id) |
            Q(parent__parent__parent__parent_id=self.location_id)
        ).values_list(
             'reportable__lower_level_outputs__indicator__programme_document__id',
             flat=True
        )
        return ProgressReport.objects.filter(programme_document_id__in=pd_ids)
```
What do You think about it Silin?

File: polymer/src/elements/ip-reporting/progress-reports-list.html [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        },
+      },
+       _isDraft: function(is_draft) {
+        return is_draft;
+      },
+
+      _linkText: function(report) {
+        if (report.is_draft) {
+          return 'Go To Draft';
+        } else if (report.progress_report_status === 'Sub') {
```
Comment (lkolacz) [inline polymer/src/elements/ip-reporting/progress-reports-list.html]: I guess, that this logic will display `Enter data` for status `Rejected`

File: django_api/django_api/apps/unicef/views.py (hunk 1 above)
Comment (n1207n) [inline django_api/django_api/apps/unicef/views.py]: Then overriding get_queryset method is right choice to do.

Comment (adamtlord) [inline django_api/django_api/apps/unicef/views.py]: That means we also need to pass location_id from the frontend in the url

File: django_api/django_api/apps/unicef/filters.py [hunk 3]
```diff
@@ -52,3 +56,36 @@ def get_location(self, queryset, name, value):
...
+    class Meta:
+        model = ProgressReport
+        fields = ['status', 'pd_ref_title', 'due_date']
+
+    def get_status(self, queryset, name, value):
+        return queryset.filter(status=value)
+
+    def get_due_status(self, queryset, name, value):
+        if value:
+            return queryset.filter(status__in=[PROGRESS_REPORT_STATUS.due, PROGRESS_REPORT_STATUS.over_due])
```
Comment (lkolacz) [inline django_api/django_api/apps/unicef/filters.py]: are You sure about it... status__in ... instead od `status=value`

File: django_api/django_api/apps/unicef/filters.py [hunk 4]
```diff
@@ -52,3 +56,36 @@ def get_location(self, queryset, name, value):
     #         return queryset.filter(
     #
     #         )
+
+
+class ProgressReportFilter(django_filters.FilterSet):
+    status = ChoiceFilter(name='status', choices=PROGRESS_REPORT_STATUS)
+    pd_ref_title = CharFilter(name='title', method='get_pd_title')
```
Comment (lkolacz) [inline django_api/django_api/apps/unicef/filters.py]: Consistent in naming -> pd_ref_title vs get_pd_title ... 
rename this please to get_pd_ref_title

File: django_api/django_api/apps/unicef/tests/test_views.py [hunk 5]
```diff
@@ -90,3 +96,72 @@ def test_detail_api(self):
...
+        url = reverse('progress-reports')
+        url += '?pd_ref_title=' + filter_string
+        response = self.client.get(url, format='json')
+
+        self.assertEquals(response.status_code, status.HTTP_200_OK)
+        self.assertEquals(len(response.data['results']), len(self.reports))
+
+    def test_list_api_filter_by_due_date(self):
+        today = datetime.datetime.today()
+        date_format = '%Y-%m-%d'
```
Comment (lkolacz) [inline django_api/django_api/apps/unicef/tests/test_views.py]: btw .. we have defined std date format ... PRINT_DATA_FORMAT .. make sure Your right!

File: django_api/django_api/apps/unicef/urls.py [hunk 6]
```diff
@@ -3,14 +3,12 @@
 from .views import (
     ProgrammeDocumentAPIView,
     ProgrammeDocumentDetailsAPIView,
+    ProgressReportAPIView,
 )
 
 
 urlpatterns = [
```
Comment (lkolacz) [inline django_api/django_api/apps/unicef/urls.py]: why You change this? this is correct PEP8 format .. remember max line should be 120!

File: polymer/src/elements/filters/date-filter/date-filter.html [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+      _handleInput: function () {
+        var newValue = this.$.field.value;
+        this.fire('filter-changed', {
+          name: this.name,
+          value: newValue,
+        });
+      },
+
+      detached: function () {
+        if (this.isDebouncerActive('input')) {
```
Comment (ErinOC) [inline polymer/src/elements/filters/date-filter/date-filter.html]: It doesn't look like you're using debounce for this component (and I'm thinking it's not necessary to add it due to the time it takes to click through the modal) so you can remove this part.

File: polymer/src/elements/ip-reporting/progress-reports-filters.html [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        </location-filter>
+
+        <date-filter
+          class="item"
+          label="Due date"
+          name="due_date"
+          value="[[_withDefault(queryParams.due_date, '')]]">
+        </date-filter>
+
+        <checkbox-filter
```
Comment (ErinOC) [inline polymer/src/elements/ip-reporting/progress-reports-filters.html]: I don't think this blocks merging, but FYI that Silin and I have been discussing changes around a similar issue on the indicators page:
The 'Report Status' dropdown would be confusing if it conflicts with 'Only show due reports' (which is basically a shortcut for selecting Due from the Report Status dropdown). So we may want to clear or disable that dropdown if the checkbox shortcut is selected.

**Pull request merged**